src/discordcr/cache.cr

```diff
@@ -1,216 +1,241 @@
-    def resolve_role(id : UInt64) : Role
-      @roles[id] # There is no endpoint for getting an individual role, so we will have to ignore that case for now.
+    # Stores the given object, expecting that `object` responds to `#id`.
+    def cache(object : V)
+      cache(object.id, object)
     end
 
-    # Resolves the ID of a DM channel with a particular user by the recipient's
-    # *recipient_id*. If there is no such channel cached, one will be created.
-    def resolve_dm_channel(recipient_id : UInt64) : UInt64
-      @dm_channels.fetch(recipient_id) do
-        channel = @client.create_dm(recipient_id)
-        cache(Channel.new(channel))
-        channel.id
-      end
-    end
+    # Removes a member from this cache from its `key`.
+    abstract def remove(key : K)
 
-    # Resolves the current user's profile. Requires no parameters since the
-    # endpoint has none either. If there is a gateway connection this should
-    # always be cached.
-    def resolve_current_user : User
-      @current_user ||= @client.get_current_user
+    # Removes a member, expecting that `object` responds to `#id`.
+    def remove(object : V)
+      remove(object.id)
     end
 
-    # Deletes a user from the cache given its *ID*.
-    def delete_user(id : UInt64)
-      @users.delete(id)
-    end
+    # Resolves the cache memeber from `key`.
+    abstract def resolve?(key : K)
 
-    # Deletes a channel from the cache given its *ID*.
-    def delete_channel(id : UInt64)
-      @channels.delete(id)
-    end
+    # Resolves the cache member from `key`.
+    # Raises an exception if the member isn't found instead of making a request.
+    def resolve(key : K)
+      member = resolve?(key)
+      raise MissingMember.new("Cache member not present with key: #{key}") unless member
 
-    # Deletes a guild from the cache given its *ID*.
-    def delete_guild(id : UInt64)
-      @guilds.delete(id)
+      member
     end
 
-    # Deletes a member from the cache given its *user_id* and the *guild_id* it
-    # is on.
-    def delete_member(guild_id : UInt64, user_id : UInt64)
-      @members[guild_id]?.try &.delete(user_id)
-    end
+    # Registers handlers on this cache's `Client` to manage its stored data
+    abstract def init_handlers
+  end
 
-    # Deletes a role from the cache given its *ID*.
-    def delete_role(id : UInt64)
-      @roles.delete(id)
-    end
+  # Container for multiple `TypeCache` objects that maps Discord types to a `TypeCache`
+  class CacheSet
+    # The `Client` to pass to the caches to register their handlers on after initialization
+    getter client : Client
+
+    def initialize(@client)
+    end
+
+    # The different possible cache types
+    TYPES = {
+      guild:        {key: UInt64, type: Guild},
+      guild_member: {key: {UInt64, UInt64}, type: GuildMember},
+      channel:      {key: UInt64, type: Channel},
+      role:         {key: {UInt64, UInt64}, type: Role},
```

Comment:
Roles are uniquely identified by their ID and don't need the guild ID like members do. Instead, there should be a separate cache type that stores guild-role associations (and others for guild-channel associations, etc.)

Reply:
The purpose of the "compound" tuple keys here is that they reflect the REST route argument types required to resolve them from the API.

Having a cache for `role: {key: UInt64, type: Role}` and something like `guild_roles: {key: UInt64, type: UInt64}` is also a valid design, but it also complicates setting up the cache / designing your own. Some may not even need this, as its only really needed if you plan on operating across the whole set of cached objects.

At the moment I've opted to leave this up to the user, and provide a second `resolve` method in the stock `Memory` caches that takes only the first ID (the guild ID) and access a second in-memory store that manages that relationship.

https://github.com/meew0/discordcr/pull/92/files/3a7578251e06aa6f28dc35d00b61e0747ea3d095#diff-a5ef42fdb346fff1caf1cb1e65c2cc16R49

Does that make sense? We can still go that way (separate relationship store).